pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -205,24 +224,59 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
 		// discovery call if the resources appeared in-between the calls. In that
 		// case, the restMapper will fail to map some of newResources until the next
 		// sync period.
-		if err := gc.resyncMonitors(newResources); err != nil {
-			utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors: %v", err))
-			return
+		if err := gc.resyncMonitors(currentResources); err != nil {
+			return map[schema.GroupVersionResource]struct{}{}, fmt.Errorf("failed to sync resource monitors: %v", err)
 		}
-		// TODO: WaitForCacheSync can block forever during normal operation. Could
-		// pass a timeout channel, but we have to consider the implications of
-		// un-pausing the GC with a partially synced graph builder.
-		if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
-			utilruntime.HandleError(fmt.Errorf("timed out waiting for dependency graph builder sync during GC sync"))
-			return
+
+		// resourcesUpdatedCh will be closed if discoveryClient returns a set of deletable resources
+		// that is different from the set that the monitors are synced to, in that case we will
+		// have to restart the process of syncing the cache.
+		resourcesUpdatedCh := make(chan struct{})
+		go func() {
+			wait.PollUntil(period, func() (bool, error) {
+				// Get the current resource list from discovery.
+				newResources := GetDeletableResources(discoveryClient)
+
+				// Decide whether discovery has reported a change.
+				if reflect.DeepEqual(currentResources, newResources) {
+					return false, nil
+				}
+
+				close(resourcesUpdatedCh)
+				return true, nil
+			}, stopCh)
```

Comment:
Does this create one go routine each time we run `resync`? We probably need to create another stop channel inside `resync`

Reply:
That is a really good point